datafusion-examples/examples/table_sample.rs

```diff
@@ -0,0 +1,1353 @@
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn properties(&self) -> &PlanProperties {
+        &self.cache
+    }
+
+    fn maintains_input_order(&self) -> Vec<bool> {
+        vec![false] // Sampling does not maintain input order
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        vec![&self.input]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(SampleExec::try_new(
+            Arc::clone(&children[0]),
+            self.lower_bound,
+            self.upper_bound,
+            self.with_replacement,
+            self.seed,
+        )?))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let input_stream = self.input.execute(partition, context)?;
+        let baseline_metrics = BaselineMetrics::new(&self.metrics, partition);
+
+        Ok(Box::pin(SampleExecStream {
+            input: input_stream,
+            sampler: self.create_sampler(partition)?,
+            baseline_metrics,
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn partition_statistics(&self, partition: Option<usize>) -> Result<Statistics> {
+        let input_stats = self.input.partition_statistics(partition)?;
+
+        // Apply sampling ratio to statistics
+        let mut stats = input_stats;
+        let ratio = self.upper_bound - self.lower_bound;
+
+        stats.num_rows = stats
+            .num_rows
+            .map(|nr| (nr as f64 * ratio) as usize)
+            .to_inexact();
+        stats.total_byte_size = stats
+            .total_byte_size
+            .map(|tb| (tb as f64 * ratio) as usize)
+            .to_inexact();
+
+        Ok(stats)
+    }
+}
+
+/// Stream for the SampleExec operator
+struct SampleExecStream {
+    /// The input stream
+    input: SendableRecordBatchStream,
+    /// The sampling method
+    sampler: Box<dyn Sampler>,
+    /// Runtime metrics recording
+    baseline_metrics: BaselineMetrics,
+}
+
+impl Stream for SampleExecStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let poll = self.input.poll_next_unpin(cx);
+        let baseline_metrics = &mut self.baseline_metrics;
+
+        match poll {
+            Poll::Ready(Some(Ok(batch))) => {
+                let start = baseline_metrics.elapsed_compute().clone();
+                let result = self.sampler.sample(&batch);
+                let _timer = start.timer();
+                Poll::Ready(Some(result))
+            }
+            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
```

Comment:
You can use the ready! macro here to make this more concise it you want

Something like
```rust
match ready!(poll) {
  Some(Ok(..)) 
...
}
```